rfcs/20210129-optimize-pd-creation/README.md

```diff
@@ -0,0 +1,262 @@
+IDs will be the following:
+
+| Implementation ID | Implementation index in the list | Valid |
+| ----------------- | -------------------------------- | ----- |
+|       -           |                0                 |   -   |
+|       0           |                1                 |   +   |
+|       -           |                2                 |   -   |
+|       1           |                3                 |   +   |
+|       -           |                4                 |   -   |
+|       2           |                5                 |   +   |
+
+In other words, the implementation IDs are the indexes in a hypothetical list
+of valid implementations for the given operation descriptors and attributes.
+
+The implementation ID can be non-zero only when using a primitive descriptor
+iterator to iterate through the list of implementations. Each attempt to iterate
+through the list to get the next valid implementation increments the implementation
+ID. In all other cases the implementation ID will be 0.
+
+Also, there are 3 primitives that do not support the iterator, those are sum,
+concat and reorder, therefore the implementation ID for those primitives is always
+0.
+
+The implementation ID will be stored in `primitive_desc_t` so that it can be
+obtained to create a key during primitive creation.
+
+#### Automatic Algorithm Selection for Convolution
+
+oneDNN supports `dnnl::algorithm::convolution_auto` algorithm that instructs
+the library to automatically select the best algorithm based on the heuristics
+that take into account tensor shapes and the number of logical processors
+available. The algorithm is specified with operation descriptor.
+
+oneDNN provides a query mechanism for getting information from oneDNN abstractions.
+For example a user can query a primitive descriptor from created primitive. oneDNN
+guarantees that the primitive descriptor that the user used to create the primitive
+is equal to the the primitive descriptor that the query mechanism returns. This is
+guaranteed not only for primitive descriptors but for all other abstractions.
+However, there is an exception - operation descriptor for convolution. The thing
+here is that if the user created primitive descriptor for convolution using
+an operation descriptor with algorithm `convolution auto` the query mechanism
+will return the operation descriptor with the algorithm that the library selected.
+This happens because oneDNN modifies the operation descriptor when creating
+primitive descriptor.
+Given that, `op_desc_t` that a user uses to create primitive descriptor and
+`op_desc_t` which is obtained from primitive descriptor to create a key to find
+the requested primitive in the cache may differ in case of convolution therefore
+they cannot be compared as is.
+
+The idea to rely on the implementation ID to distinguish the algorithm and ignore
+the field in operation descriptor during comparison.
```

Comment:
This places a constraint on our dispatching: all algorithms shall be dispatched from the same list.
Are we ok with that? @kwiersch what do you think?

Reply:
@mgouicem As far as I understand, there is currently no difference in implementation lists for `auto`, `direct` and `wino` algorithms of convolution. So from that perspective the constraint is already satisfied. Looking forward, I think it would be hard to disentangle `direct` and `wino` since we have `auto` to choose between the two, so I have no objections.